### PR TITLE
[IMP] crm_*: add more fields when we merge multiple leads

### PR DESCRIPTION
**Pull request by std-odoo**
Purpose
=======
Add more fields when we merge multiple leads to be sure not to lose
valuable information.

Specifications
==============
New fields merged
-----------------
Those fields are now propagated to the destination,
- referred
- color
- recurring_revenue
- recurring_plan
- function
- lang_id
- date_deadline
- reveal_id
- lead_mining_request_id
- reveal_ip
- reveal_iap_credits
- reveal_rule_id
- event_lead_rule_id
- event_id
- lost_reason (if the destination lead is lost)
- iap_enrich_done
- tag_ids

The address is taken from the lead with the most non-empty address
fields (sorted by highest rank if multiple leads have the same amount
of non-empty fields).

Followers
---------
When merging multiple leads, we want to move the followers to the
destination lead if they posted a message in the last 30 days.

A message will be added in the merge note to know which followers have
been added.

Views
-----
Remove the "groups" attribute  on the recurring revenue
fields. Instead we hide them directly in the views.

Task-2447721

### CODE REVIEW

File: addons/crm/models/crm_lead.py [hunk 1]
```diff
@@ -1303,12 +1303,26 @@ def _merge_opportunity(self, user_id=False, team_id=False, auto_unlink=True, max
...
             'type': lambda fname, leads: 'opportunity' if any(lead.type == 'opportunity' for lead in leads) else 'lead',
             'priority': lambda fname, leads: max(leads.mapped('priority')) if leads else False,
+            'tag_ids': lambda fname, leads: leads.mapped('tag_ids'),
+            'lost_reason': lambda fname, leads:
+                False if leads and leads[0].probability
+                else next((lead.lost_reason for lead in leads if lead.lost_reason), False),
         }
 
+        # Address related fields, all fields are propagated or none
+        fields_specific.update({
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:1320]: Looking at it, it seems it is not really all fields or none, more "first found containing any field". Shouldn't we try to get a bit more information  (like country on first lead, then full address on second lead) ?

File: addons/crm/models/crm_lead.py [hunk 2]
```diff
@@ -1303,12 +1303,26 @@ def _merge_opportunity(self, user_id=False, team_id=False, auto_unlink=True, max
...
+            field: lambda fname, leads: next((
+                lead[fname] for lead in leads
+                if any(lead[field] for field in PARTNER_ADDRESS_FIELDS_TO_SYNC)), False)
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:1323]: Actually it is a bit tricky, thinking of it. Do we want to accept an address from an higher ranker lead, or a full address from a lower ranked lead ? Would day that we could fetch address information of all merged leads, and take the one with the most information (taking higher ranked lead if equivalent number of fields given).

File: addons/crm/tests/common.py [hunk 3]
```diff
@@ -417,13 +417,21 @@ def _aggregate(fname):
...
+                    return lead[fname]
+
         try:
             # merge process will modify opportunity
             yield
         finally:
             # support specific values caller may want to check in addition to generic tests
             for fname, expected in expected.items():
-                self.assertEqual(opportunity[fname], expected)
+                if expected is False:
```
Comment (tde-banana-odoo) [inline addons/crm/tests/common.py:431]: OOC, why this change ?

File: addons/crm/tests/common.py [hunk 4]
```diff
@@ -616,6 +631,11 @@ def setUpClass(cls):
             'partner_id': cls.contact_company_1.id,
             'contact_name': 'Hermes Conrad',
             'email_from': 'hermes.conrad@test.example.com',
+            'tag_ids': [
+                (4, cls.tags[0].id),
+                (4, cls.tags[1].id),
+                (0, 0, {'name': 'Tag 5'}),
```
Comment (tde-banana-odoo) [inline addons/crm/tests/common.py:637]: Do we need that much data ? In any way, would create all tags first, then give them to sub records (0-3 to lead_1, 0-1-4 to lead_w_partner_company, ...). That way we can always compare records based on created data.

File: addons/crm/models/crm_lead.py [hunk 5]
```diff
@@ -1405,6 +1408,51 @@ def _merge_dependences_calendar_events(self, opportunities):
...
+        # leads in the last 7 days) which should be moved on the destination lead
+        self.env.cr.execute(
+            '''
+            SELECT MAX(mf.id) AS id
+              FROM mail_followers AS mf
+              JOIN mail_message AS mm
+                ON mm.author_id = mf.partner_id
+               AND mm.res_id = mf.res_id
+               AND mm.model = 'crm.lead'
+               AND mm.date > NOW() - INTERVAL '7 DAY'
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:1433]: Thinking of it, if we add message authors only (which is quite restrictive), we could probably increase the interval to two months or something like that. Indeed if we merge a lead staging since a few times, probably is high that there was no message in the last few days. What do you think ?

**Review by tde-banana-odoo — COMMENTED** [review]
Validation ! Added some comments and questions, for address and followers notably. Did not look in details for tests, will do it once other points are validated.

Thanks !

File: addons/crm/models/crm_lead.py (hunk 1 above)
Comment (std-odoo) [reply addons/crm/models/crm_lead.py:1327]: I did your idea bellow :) (address with the most non-empty fields)

File: addons/crm/tests/common.py (hunk 3 above)
Comment (std-odoo) [reply addons/crm/tests/common.py:431]: So I can check that a field is falsy (in that case it might be `False` or `self.env['res.country']`
```
self.assertLeadMerged(self.lead_w_contact, leads, country_id=False):
```

File: addons/crm/models/crm_lead.py (hunk 5 above)
Comment (std-odoo) [reply addons/crm/models/crm_lead.py]: Do not really know in practice, maybe it's safer 2 months, I make the change and will notify LNA

**Review by std-odoo — COMMENTED** [review]
Hi :) I did the changes and answered :)

File: addons/crm/tests/common.py [hunk 6]
```diff
@@ -417,13 +417,21 @@ def _aggregate(fname):
                 values = [_find_value(lead, fname) for lead in leads]
             return values
 
+        def _get_address(fname):
+            for lead in leads:
```
Comment (tde-banana-odoo) [inline addons/crm/tests/common.py:421]: Shouldn't we explicitly give the "address valid" lead in assert to ensure the right one is chosen ? I fear we may have bugs hidden in an heuristic in a test ...

File: addons/crm/tests/test_crm_lead_merge.py [hunk 7]
```diff
@@ -84,6 +85,71 @@ def test_initial_data(self):
...
+            'street': 'Test street',
+            'street2': 'Test street2',
+            'city': 'Test City',
+            'zip': '5000',
+            'state_id': False,
+            'country_id': self.env.ref('base.be'),
+        }
+        self.lead_w_contact.write(initial_address)
+
+        # 5 non-empty address field like for the destination field, should be ignored
```
Comment (tde-banana-odoo) [inline addons/crm/tests/test_crm_lead_merge.py:102]: They are ignored because we have "5 data (like initial address) but less prior", is it right ? Could gain a comment by the way :D

File: addons/crm/tests/test_crm_lead_merge.py [hunk 8]
```diff
@@ -294,3 +390,120 @@ def test_merge_method_dependencies(self):
...
+        self.leads.mapped('message_follower_ids').unlink()
+        self.leads.mapped('message_ids').unlink()
+
+        (
+            follower_contact_1_lead_1,
+            follower_contact_1_lead_2,
+            follower_contact_2_lead_2,
+            follower_contact_2_lead_3,
+            follower_contact_3_lead_2,
+        ) = self.env['mail.followers'].sudo().create([{
```
Comment (tde-banana-odoo) [inline addons/crm/tests/test_crm_lead_merge.py:413]: Would prefer using a standard message_subscribe, to avoid low-level manipulation of followers.

File: addons/crm/tests/test_crm_lead_merge.py [hunk 9]
```diff
@@ -294,3 +390,120 @@ def test_merge_method_dependencies(self):
...
+    @users('user_sales_manager')
+    @mute_logger('odoo.models.unlink')
+    def test_merge_method_followers(self):
+        """ Test that the followers of the leads are added in the destination lead.
+
+        They should be added if:
+        - The related partner was active on the lead (posted a message in the last 7 days)
+        - The related partner is not already following the destination lead
+        """
+        self.env['crm.lead'].browse(self.lead_w_partner_company.ids).write({'type': 'opportunity'})
```
Comment (tde-banana-odoo) [inline addons/crm/tests/test_crm_lead_merge.py:403]: Why this change ? Can't we rely on standard test data ? We already have some tests about confidence computation normally. That way we keep scope of the test only for followers.

File: addons/crm/tests/test_crm_lead_merge.py [hunk 10]
```diff
@@ -294,3 +390,120 @@ def test_merge_method_dependencies(self):
...
+        self.leads.mapped('message_follower_ids').unlink()
```
Comment (tde-banana-odoo) [inline addons/crm/tests/test_crm_lead_merge.py:404]: Mapped is probably not necessary (same below)

File: addons/event_crm/views/crm_lead_merge_template.xml [hunk 11]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+<template id="crm_lead_merge_summary_inherit_event_crm" inherit_id="crm.crm_lead_merge_summary">
+    <xpath expr="//div[@name='marketing']" position="after">
+        <div class="mt-3 mb-3" name="event"
+            t-if="lead.event_lead_rule_id or lead.event_id">
+        <div class="font-weight-bold">
```
Comment (tde-banana-odoo) [inline addons/event_crm/views/crm_lead_merge_template.xml:8]: Missing indent ?

File: addons/crm/tests/test_crm_lead_merge.py [hunk 12]
```diff
@@ -294,3 +390,120 @@ def test_merge_method_dependencies(self):
...
+
+        # Check followers of the destination lead
+        new_followers = master_lead.message_follower_ids - initial_followers
+        self.assertIn(follower_contact_2_lead_3, new_followers,
+                      'The partner must be followed the destination lead')
+        self.assertNotIn(follower_contact_1_lead_2, new_followers,
+                         'The partner was already followed the destination lead')
+        self.assertNotIn(follower_contact_2_lead_2, new_followers,
+                         'The partner was already added')
+        # "contact_company" posted a message 8 days ago on lead_2, so it's considered as inactive
```
Comment (tde-banana-odoo) [inline addons/crm/tests/test_crm_lead_merge.py:501]: Would be great to have a docstring at the beginning of the test, like a small table / summary telling what are test data and what should be transfered to the new lead so that we could quickly understand what is going on in case of issue. Here it is a bit split between data and comments. Something like
```
Lead                    - Contact     - Info
lead_w_partner (MASTER) - contact_2   - OK (author message_2)
lead_w_partner (MASTER) - contact_4   - OK (author message_4)
lead_32                 - contact_3   - KO (author message_3 but old)
lead_32                 - contact_5   - KO (not author)
...
```

**Review by tde-banana-odoo — COMMENTED** [review]
Second look, business code LGTM, some comments for tests :)

File: addons/crm/tests/common.py (hunk 3 above)
Comment (std-odoo) [reply addons/crm/tests/common.py:431]: yes :)

File: addons/crm/tests/common.py (hunk 6 above)
Comment (std-odoo) [reply addons/crm/tests/common.py:421]: We can just ignore the address field and let us give the value manually (like we already do)

File: addons/event_crm/views/crm_lead_merge_template.xml (hunk 11 above)
Comment (std-odoo) [reply addons/event_crm/views/crm_lead_merge_template.xml:8]: Good catch :)

File: addons/crm/tests/common.py (hunk 6 above)
Comment (tde-banana-odoo) [reply addons/crm/tests/common.py:421]: Probably easier indeed :)

File: addons/crm/tests/test_crm_lead_merge.py (hunk 12 above)
Comment (std-odoo) [reply addons/crm/tests/test_crm_lead_merge.py:501]: good idea :)

**Review by awa-odoo — COMMENTED** [review]
Hello @std-odoo 

Nice task with a lot of cleaning and some good logic.
This will make the merge message look all shiny with the latest additions from NEB.

I added a bit of everything, some polish, some questions, some logic simplifying.
Nothing major though so we should be good soon.

Thanks for your work!

**Review by awa-odoo — COMMENTED** [review]
A couple more minor things and then we are all good!
